### PR TITLE
[QC][PPU] Add fused_inv_rope_fp8_quant operator

### PR DESCRIPTION
**Pull request by July-h5kf3**
### PR Category

Operator

### Type of Change

New Feature, Performance Optimization

### Description

Adds PPU (`thead`) implementations of `fused_inv_rope_fp8_quant` and `per_token_group_quant_fp8`, registered through the existing vendor override mechanism. All implementation changes are under `runtime/backend/_thead/`; generic ops, tests, and benchmark files are unchanged.

FlagTree PPU cannot lower native `tl.float8e4nv` pointers or conversions. Both operators encode E4M3FN bytes with integer operations and store through a no-copy uint8 view of the FP8 output allocation. Quantization remains FP8 E4M3FN with FP32 or packed UE8M0 scales as appropriate; there is no torch compute fallback.

`per_token_group_quant_fp8` uses one group per program and compile-time group dimensions, FP8 bounds, and scale mode. This implementation takes 2.80 us for `(7, 4096, 256)` with standard scales. Row-major and column-major scale output remain supported.

The fused inverse-RoPE kernel retains the narrow RoPE tile, static token-block dispatch, integer power-of-two scale calculation, and head-first launch grid described in the original implementation.

### Validation

PPU-ZW810E device 3, Python 3.12, PyTorch 2.10.0, FlagTree Triton 3.6.0; commit `2bdcacda01debde5a9983f8cd33951a40389cc8d`:

- `PYTHONPATH=src python -m pytest -q tests/test_per_token_group_quant_fp8.py`: **96 passed**.
- Additional standalone checks: **72 passed** across FP16/BF16/FP32, group sizes 64/256/512, contiguous and strided rows, row/column-major scales, standard/UE8M0 scales, zero groups and signed zero.
- Both benchmark implementations checked against a CPU FP8 reference before timing all ten cases.
- `pre-commit run --all-files`: **all hooks passed**. Production-path audit found only tensor allocation, metadata queries and no-copy views; kernel casts are Triton operations.
- Current `tests/test_fused_inv_rope_fp8_quant.py`: **124 skipped** by the existing native-FP8 gate. The earlier 124-pass record predates the generic-op cleanup and is not a validation claim for this revision.

### per_token_group_quant_fp8 performance

BF16 inputs, FP8 E4M3FN output, FP32 row-major scales, fixed seed 0. Shapes come directly from this PR's `benchmark/core_shapes.yaml`; each is measured with standard and UE8M0 scales. Timing uses `triton.testing.do_bench(warmup=25, rep=100, return_mode="median")`, with three alternating-order rounds and the median of those rounds. Latencies are microseconds; speedup is baseline latency divided by this implementation's latency. Device 3 was idle before the run.

**Baseline limitation:** the installed PPU vLLM distribution (`0.24.0+empty`) cannot import because `_C_stable_libtorch` is missing. Its native FP8 Triton kernel also fails to compile with `type fp8e4nv not supported in this architecture`. The comparison therefore uses the vLLM Triton fallback with only the FP8 conversion replaced by the same manual E4M3FN encoder and an output uint8 view. It is an **adapted vLLM Triton baseline**, not the installed/native vLLM CUDA operator. The group reduction, scaling, indexing, one-group grid, and launch configuration are preserved. The extracted kernel matches [upstream vLLM source](https://github.com/vllm-project/vllm/blob/59f15272802077724c4ed483ea8ad393725554e4/vllm/model_executor/layers/quantization/utils/fp8_utils.py).

| Shape (tokens, width, group) | Scale | Adapted vLLM FP8 (us) | This implementation (us) | Speedup |
|---|---|---:|---:|---:|
| `(7, 512, 512)` | standard | 3.48 | 3.00 | 1.160x |
| `(7, 4096, 256)` | standard | 3.04 | 2.80 | 1.086x |
| `(83, 512, 64)` | standard | 3.04 | 2.80 | 1.086x |
| `(2048, 4096, 256)` | standard | 27.12 | 25.68 | 1.056x |
| `(2048, 13824, 512)` | standard | 89.40 | 84.32 | 1.060x |
| `(7, 512, 512)` | UE8M0 | 3.72 | 3.24 | 1.148x |
| `(7, 4096, 256)` | UE8M0 | 3.48 | 3.00 | 1.160x |
| `(83, 512, 64)` | UE8M0 | 3.04 | 2.80 | 1.086x |
| `(2048, 4096, 256)` | UE8M0 | 30.20 | 27.80 | 1.086x |
| `(2048, 13824, 512)` | UE8M0 | 99.44 | 92.20 | 1.079x |

Arithmetic mean of the ten unrounded speedups: **1.101x**. Minimum speedup: **1.056x**. These figures satisfy the local same-precision target against this adapted baseline; no claim is made about an unavailable native vLLM implementation.

<details>
<summary>Standalone PPU reproduction script</summary>

Run inside the existing PPU test environment from the PR checkout. Select an idle card first. The script contains the adapted baseline, uses the checkout's benchmark shapes, validates both outputs, and writes all rounds to `results.json` beside the script.

```bash
CUDA_VISIBLE_DEVICES=3 PYTHONPATH=src /opt/venv/sdk21/bin/python -m pytest -q tests/test_per_token_group_quant_fp8.py
# Save the Python block below as /tmp/reproduce_per_token.py, then run:
CUDA_VISIBLE_DEVICES=3 PYTHONPATH=src /opt/venv/sdk21/bin/python /tmp/reproduce_per_token.py
```

```python
import hashlib
import importlib.metadata
import json
import os
from pathlib import Path
import statistics
import sys

import torch
import triton
import triton.testing
import yaml

import flaggems_vllm


import triton
import triton.language as tl
from flaggems_vllm.runtime.backend._thead.ops.per_token_group_quant_fp8 import _float_to_e4m3fn_bits

@triton.jit
def _per_token_group_quant_fp8(y_ptr, y_q_ptr, y_s_ptr, group_size, y_num_columns, y_row_stride, eps, fp8_min: tl.constexpr, fp8_max: tl.constexpr, use_ue8m0: tl.constexpr, BLOCK: tl.constexpr):
    """A Triton-accelerated function to perform per-token-group
    quantization on a tensor.
    This function converts the tensor values into float8 values.
    """
    groups_per_row = y_num_columns // group_size
    g_id = tl.program_id(0)
    row = g_id // groups_per_row
    row_g_id = g_id % groups_per_row
    y_ptr_offset = row.to(tl.int64) * y_row_stride + row_g_id.to(tl.int64) * group_size
    y_ptr += y_ptr_offset
    y_q_ptr_offset = g_id.to(tl.int64) * group_size
    y_q_ptr += y_q_ptr_offset
    y_s_ptr += g_id
    cols = tl.arange(0, BLOCK)
    mask = cols < group_size
    y = tl.load(y_ptr + cols, mask=mask, other=0.0).to(tl.float32)
    _absmax = tl.maximum(tl.max(tl.abs(y)), eps)
    scale_raw = _absmax * (1.0 / fp8_max)
    y_s = tl.math.exp2(tl.ceil(tl.log2(scale_raw))) if use_ue8m0 else scale_raw
    y_q = _float_to_e4m3fn_bits(tl.clamp(y / y_s, fp8_min, fp8_max)).to(tl.uint8)
    tl.store(y_q_ptr + cols, y_q, mask=mask)
    tl.store(y_s_ptr, y_s)


def baseline(x, group_size, use_ue8m0):
    quantized = torch.empty(x.shape, device=x.device, dtype=torch.float8_e4m3fn)
    scales = torch.empty(
        (x.shape[0], x.shape[1] // group_size), device=x.device, dtype=torch.float32
    )
    block = triton.next_power_of_2(group_size)
    _per_token_group_quant_fp8[(x.numel() // group_size,)](
        x, quantized.view(torch.uint8), scales, group_size, x.shape[1],
        x.stride(0), 1e-10, fp8_min=-448.0, fp8_max=448.0,
        use_ue8m0=use_ue8m0, BLOCK=block,
        num_warps=min(max(block // 256, 1), 8), num_stages=1,
    )
    return quantized, scales


def main():
    torch.manual_seed(0)
    root = Path(__file__).parent
    repo = Path(flaggems_vllm.__file__).parents[2]
    shapes = yaml.safe_load((repo / 'benchmark/core_shapes.yaml').read_text())[
        'PerTokenGroupQuantFp8Benchmark'
    ]['shapes']
    metadata = {
        'python': sys.version, 'torch': torch.__version__, 'triton': triton.__version__,
        'vllm_distribution': importlib.metadata.version('vllm'),
        'device': torch.cuda.get_device_name(0),
        'CUDA_VISIBLE_DEVICES': os.environ['CUDA_VISIBLE_DEVICES'],
        'operator_module': flaggems_vllm.per_token_group_quant_fp8.__module__,
        'vllm_source_sha256': hashlib.sha256(Path('/opt/vllm/vllm/model_executor/layers/quantization/utils/fp8_utils.py').read_bytes()).hexdigest(),
        'timing': 'triton.testing.do_bench; warmup=25ms, rep=100ms, median; 3 alternating-order rounds',
        'baseline': 'installed vLLM Triton fallback with manual E4M3FN conversion and uint8 output view',
        'rows': [],
    }
    for ue8m0 in (False, True):
        for tokens, width, group_size in shapes:
            x = torch.rand((tokens, width), device='cuda', dtype=torch.bfloat16)
            reference_input = x.cpu().float().reshape(-1, group_size)
            scales = reference_input.abs().amax(dim=1, keepdim=True).clamp(min=1e-10) / 448
            if ue8m0:
                scales = torch.exp2(torch.ceil(torch.log2(scales.clamp(min=1e-10))))
            quantized = (reference_input / scales).clamp(-448, 448).to(torch.float8_e4m3fn).float()
            calls = {
                'vllm': lambda: baseline(x, group_size, ue8m0),
                'gems': lambda: flaggems_vllm.per_token_group_quant_fp8(x, group_size, scale_ue8m0=ue8m0),
            }
            for call in calls.values():
                q, s = call()
                torch.testing.assert_close(s.cpu().reshape(-1, 1), scales, rtol=1e-6, atol=1e-10)
                torch.testing.assert_close(q.cpu().float().reshape(-1, group_size), quantized, rtol=0.125, atol=0.001953125)
            timings = {'vllm': [], 'gems': []}
            for round_index in range(3):
                for name in (('vllm', 'gems') if round_index % 2 == 0 else ('gems', 'vllm')):
                    timings[name].append(1000 * triton.testing.do_bench(calls[name], warmup=25, rep=100, return_mode='median'))
            vllm_us = statistics.median(timings['vllm'])
            gems_us = statistics.median(timings['gems'])
            row = {'shape': [tokens, width, group_size], 'ue8m0': ue8m0, 'vllm_us': vllm_us,
                   'gems_us': gems_us, 'speedup': vllm_us / gems_us, 'rounds_us': timings}
            metadata['rows'].append(row)
            print(json.dumps(row), flush=True)
    metadata['mean_speedup'] = statistics.mean(r['speedup'] for r in metadata['rows'])
    (root / 'results.json').write_text(json.dumps(metadata, indent=2) + '\n')
    print('mean speedup:', metadata['mean_speedup'], flush=True)


if __name__ == '__main__':
    main()
```

</details>

### Earlier fused_inv_rope_fp8_quant measurements

The following fused-kernel measurements were recorded during the original PPU implementation, before the final generic-op cleanup. This per-token follow-up does not re-measure or revalidate these historical fused-kernel numbers. The standalone baseline used the vLLM fused kernel with the same PPU manual-FP8 adaptation.

Upstream benchmark shapes `(1, 8, 1, True)` and `(16, 64, 8, True)` (num_tokens, num_heads, n_groups, tma_aligned_scales):

|Shape|vLLM fp8 baseline us|this op us|SpeedUp|
|---|---:|---:|---:|
|`(1, 8, 1, True)`|5.60|4.20|1.333|
|`(16, 64, 8, True)`|7.00|5.60|1.250|

Token sweep (num_tokens, num_heads, n_groups):

|Shape|vLLM fp8 baseline us|this op us|SpeedUp|
|---|---:|---:|---:|
|`(128, 128, 8)`|46.01|25.15|1.829|
|`(512, 64, 8)`|87.63|45.38|1.931|
|`(512, 128, 8)`|159.61|80.46|1.984|
|`(1024, 64, 8)`|159.14|79.09|2.012|
|`(1024, 128, 8)`|301.43|140.34|2.148|
|`(4096, 64, 8)`|591.80|251.56|2.353|
|`(4096, 128, 8)`|1164.44|476.12|2.446|
|`(8192, 64, 8)`|1185.78|475.44|2.494|
|`(8192, 128, 8)`|2347.24|919.34|2.553|

Arithmetic mean over the full 24-shape set (the two tables above plus the remaining `(heads, groups)` sweeps): **1.79x vs the vLLM fp8 baseline**, with every measured shape >= 1.0x and a minimum of ~1.0 on the single-token smoke shape.

### Issue

None.

### Progress

- [ ] Change is properly reviewed.
- [x] Per-token quantization is covered by correctness checks and benchmark results.
- [ ] Current fused-operator tests execute on PPU (currently skipped by the native-FP8 gate).

### CODE REVIEW

File: src/flaggems_vllm/ops/fused_inv_rope_fp8_quant.py [hunk 1]
```diff
@@ -19,10 +19,15 @@
...
 
+from flaggems_vllm import runtime
 from flaggems_vllm.runtime import torch_device_fn
 from flaggems_vllm.utils.device_info import get_device_capability
 
-if torch_device_fn.is_available() and get_device_capability() >= (9, 0):
+# NVIDIA SM90+ has native FP8; on PPU (thead) the hardware casts work and the
+# device overrides the Triton kernel with a manual-conversion variant below.
+if torch_device_fn.is_available() and (
+    get_device_capability() >= (9, 0) or runtime.device.vendor_name == "thead"
```
Comment (0x45f) [inline src/flaggems_vllm/ops/fused_inv_rope_fp8_quant.py]: Why is it necessary to make a special case for the `thead` here?

File: src/flaggems_vllm/ops/per_token_group_quant_fp8.py [hunk 2]
```diff
@@ -19,18 +19,53 @@
...
 
+from flaggems_vllm import runtime
 from flaggems_vllm.runtime import torch_device_fn
 from flaggems_vllm.utils.device_info import get_device_capability
 
-if torch_device_fn.is_available() and get_device_capability() >= (9, 0):
+# NVIDIA SM90+ has native FP8; on PPU (thead) the torch-side FP8 casts work
+# and the device overrides the Triton kernel with a manual-conversion variant.
+if torch_device_fn.is_available() and (
+    get_device_capability() >= (9, 0) or runtime.device.vendor_name == "thead"
```
Comment (0x45f) [inline src/flaggems_vllm/ops/per_token_group_quant_fp8.py]: Why is it necessary to make a special case for the thead here?

**Review by 0x45f — COMMENTED** [review]
why modify ‎src/flaggems_vllm/ops/xx files

File: src/flaggems_vllm/ops/per_token_group_quant_fp8.py (hunk 2 above)
Comment (July-h5kf3) [reply src/flaggems_vllm/ops/per_token_group_quant_fp8.py]: Done — generic ops/ reverted; everything PPU-specific is now fully vendored under runtime/backend/_thead/, NVIDIA paths are untouched.

**Comment by 0x45f** [comment]
please show per_token_group_quant_fp8 performance

**Comment by July-h5kf3** [comment]
> please show per_token_group_quant_fp8 performance

Done

**Review by 0x45f — APPROVED** [review]
